include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -0,0 +1,661 @@
+             ExcColMapMissmatch());
+      Assert(dst.trilinos_rcp()->getMap()->isSameAs(*matrix->getRangeMap()),
+             ExcDomainMapMissmatch());
+      matrix->apply(*src.trilinos_rcp(), *dst.trilinos_rcp());
+    }
+
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::Tvmult(Vector<Number>       &dst,
+                                       const Vector<Number> &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+      Assert(dst.trilinos_rcp()->getMap()->isSameAs(*matrix->getDomainMap()),
+             ExcColMapMissmatch());
+      Assert(src.trilinos_rcp()->getMap()->isSameAs(*matrix->getRangeMap()),
+             ExcDomainMapMissmatch());
+      matrix->apply(*src.trilinos_rcp(), *dst.trilinos_rcp(), Teuchos::TRANS);
+    }
+
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::vmult_add(Vector<Number>       &dst,
+                                          const Vector<Number> &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+      Assert(src.trilinos_rcp()->getMap()->isSameAs(*matrix->getDomainMap()),
+             ExcColMapMissmatch());
+      Assert(dst.trilinos_rcp()->getMap()->isSameAs(*matrix->getRangeMap()),
+             ExcDomainMapMissmatch());
+      matrix->apply(*src.trilinos_rcp(),
+                    *dst.trilinos_rcp(),
+                    Teuchos::NO_TRANS,
+                    Teuchos::ScalarTraits<Number>::one(),
+                    Teuchos::ScalarTraits<Number>::one());
+    }
+
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::Tvmult_add(Vector<Number>       &dst,
+                                           const Vector<Number> &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+      Assert(dst.trilinos_rcp()->getMap()->isSameAs(*matrix->getDomainMap()),
+             ExcColMapMissmatch());
+      Assert(src.trilinos_rcp()->getMap()->isSameAs(*matrix->getRangeMap()),
+             ExcDomainMapMissmatch());
+      matrix->apply(*src.trilinos_rcp(),
+                    *dst.trilinos_rcp(),
+                    Teuchos::NO_TRANS,
+                    Teuchos::ScalarTraits<Number>::one(),
+                    Teuchos::ScalarTraits<Number>::one());
+    }
+
+
+    template <typename Number, typename Node>
+    void
+    SparseMatrix<Number, Node>::print(
+      std::ostream &out,
+      const bool    print_detailed_trilinos_information) const
+    {
+      if (print_detailed_trilinos_information)
+        {
+          auto teuchos_out = Teuchos::getFancyOStream(Teuchos::rcpFromRef(out));
+          matrix->describe(*teuchos_out, Teuchos::VERB_EXTREME);
```

Comment:
Ha, yes, sure, if one wants *detailed* information, then `EXTREME` seems right :-)

Reply:
We have to use `EXTREME` here. Otherwise, Tpetra will not print the matrix itself.